### ISSUE TITLE
MSBuild v14 compiles a semantically incorrect assembly under some rare circumstances

### ISSUE THREAD

**Issue by dtereanszky** [dotnet/roslyn#11576]
**Version Used**: 
MSBuild v14

**Steps to Reproduce**:
```
static void Main(string[] args)
{
    var customerId = Guid.NewGuid();
    TestMethodWithParams(args: customerId, whatever: "foo");
}

private static void TestMethodWithParams(string whatever, params object[] args)
{
    Console.WriteLine("args: '{0}'", args);
}
```

**Expected Behavior**:
Correct binary is generated.

**Actual Behavior**:
Buggy binary is generated, parameter gets lost during method call. This can be verified by decompiling the binary:
```
Guid guid = Guid.NewGuid();
object obj;
Program.TestMethodWithParams("foo", new object[]
{
    obj // <- this is and will always be null
});
```

Please see more details here: http://stackoverflow.com/questions/37457082/msbuild-v14-compiles-a-semantically-incorrect-assembly-under-some-rare-circumsta

**Comment by WiredUK** [comment]
A simple test using the above code, just switch the order of the parameters and you get different output:

    TestMethodWithParams(args: customerId, whatever: "foo");
    TestMethodWithParams(whatever: "foo", args: customerId);

**Comment by dtereanszky** [comment]
Hi WiredUk,

thank you for your suggestion. It works, just as wrapping that very single parameter into an object array:
TestMethodWithParams(args: **new object[] {**customerId**}**, whatever: "foo");

Anyway, these are good as workaround, but currently seems this behaviour is a bug, that should probably be fixed in my opinion.

**Comment by svick** [comment]
I believe this is a duplicate of #4197, which was fixed in 1.1.

**Comment by dtereanszky** [comment]
Hi @svick,

installing "Microsoft Build Tools 2015 with Update 2" solved our problem.

Thank you.

/David
